{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f3dbf9cd-8440-42ff-be9d-51f9722dac34",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil, subprocess, time\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "pd.set_option('display.max_columns', None)\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import itertools\n",
    "%matplotlib inline\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.dds import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ace384e-bdcb-43dc-ba0a-5b962fa4bf55",
   "metadata": {},
   "source": [
    "### Verification sgRNA ABE Screen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ff16de91-8696-4260-b806-4c54f37406bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/media/ivan/Extreme SSD/Verification ABE Screen')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "58e2c8e5-0294-463f-82f0-75d4cf8e0b9a",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid decimal literal (2906377291.py, line 5)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[3], line 5\u001b[0;36m\u001b[0m\n\u001b[0;31m    cutadapt -g TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAATAGCAAGTTA -j 10 -m 18 -M 23 I75A714_CKDL220019605-1A_HHT7GCCX2_L3_1.fq -o test_cut2.fq\u001b[0m\n\u001b[0m                                                                                                        ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid decimal literal\n"
     ]
    }
   ],
   "source": [
    "#in terminal\n",
    "gunzip *.gz\n",
    "\n",
    "\n",
    "cutadapt -g TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAATAGCAAGTTA -j 10 -m 18 -M 23 I75A714_CKDL220019605-1A_HHT7GCCX2_L3_1.fq -o test_cut2.fq\n",
    "\n",
    "bowtie-build Verification_sgRNA.fasta Veri\n",
    "bowtie -n 2 -p 20 -q --best --norc -x BCR_ABL_Control_sgRNA test_cut2.fq > results.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ae1a0c63-0360-4d21-a302-5fa86ef23ca2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 3.5 with Python 3.10.12\n",
      "Command line parameters: -g TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT -j 20 -m 17 -M 23 Veri_L1713.fq -o Trimmed/Veri_L1713.fq\n",
      "Processing reads on 20 cores in single-end mode ...\n",
      "Finished in 0.22 s (3 µs/read; 19.27 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  72,185\n",
      "Reads with adapters:                    68,634 (95.1%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                 143 (0.2%)\n",
      "Reads that were too long:               16,041 (22.2%)\n",
      "Reads written (passing filters):        56,001 (77.6%)\n",
      "\n",
      "Total basepairs processed:    10,827,750 bp\n",
      "Total written (filtered):      1,131,603 bp (10.5%)\n",
      "\n",
      "=== Adapter 3 ===\n",
      "\n",
      "Sequence: TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT; Type: linked; Length: 18+17; 5' trimmed: 68634 times; 3' trimmed: 68634 times\n",
      "\n",
      "Minimum overlap: 3+3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-18 bp: 1\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-17 bp: 1\n",
      "\n",
      "Overview of removed sequences at 5' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "10\t2\t0.1\t1\t2\n",
      "14\t1\t0.0\t1\t0 1\n",
      "16\t1\t0.0\t1\t0 1\n",
      "17\t2\t0.0\t1\t0 1 1\n",
      "18\t2\t0.0\t1\t2\n",
      "19\t4\t0.0\t1\t4\n",
      "20\t68\t0.0\t1\t10 54 4\n",
      "21\t8182\t0.0\t1\t7158 981 43\n",
      "22\t8258\t0.0\t1\t7161 1035 62\n",
      "23\t16781\t0.0\t1\t14576 2112 93\n",
      "24\t13783\t0.0\t1\t12063 1662 58\n",
      "25\t21532\t0.0\t1\t19060 2397 75\n",
      "26\t18\t0.0\t1\t3 12 3\n",
      "\n",
      "\n",
      "\n",
      "Overview of removed sequences at 3' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t4\t1127.9\t0\t4\n",
      "5\t1\t70.5\t0\t1\n",
      "24\t2\t0.0\t1\t2\n",
      "25\t4\t0.0\t1\t1 3\n",
      "26\t2\t0.0\t1\t1 1\n",
      "27\t6\t0.0\t1\t6\n",
      "29\t3\t0.0\t1\t3\n",
      "41\t1\t0.0\t1\t1\n",
      "48\t1\t0.0\t1\t1\n",
      "49\t4\t0.0\t1\t3 1\n",
      "50\t3\t0.0\t1\t2 0 1\n",
      "51\t1\t0.0\t1\t1\n",
      "62\t1\t0.0\t1\t0 1\n",
      "63\t189\t0.0\t1\t165 24\n",
      "64\t1271\t0.0\t1\t1089 178 4\n",
      "65\t3233\t0.0\t1\t2746 470 17\n",
      "66\t2634\t0.0\t1\t2223 394 17\n",
      "67\t2492\t0.0\t1\t2094 377 21\n",
      "68\t1487\t0.0\t1\t1249 226 12\n",
      "69\t944\t0.0\t1\t809 130 5\n",
      "70\t27\t0.0\t1\t16 11\n",
      "71\t8\t0.0\t1\t6 2\n",
      "72\t5\t0.0\t1\t4 1\n",
      "73\t7\t0.0\t1\t6 1\n",
      "74\t8\t0.0\t1\t7 1\n",
      "75\t5\t0.0\t1\t3 2\n",
      "76\t7\t0.0\t1\t5 2\n",
      "77\t4\t0.0\t1\t4\n",
      "78\t6\t0.0\t1\t5 1\n",
      "79\t1\t0.0\t1\t1\n",
      "80\t1\t0.0\t1\t1\n",
      "81\t5\t0.0\t1\t4 1\n",
      "82\t2\t0.0\t1\t2\n",
      "83\t3\t0.0\t1\t3\n",
      "84\t7\t0.0\t1\t6 1\n",
      "85\t8\t0.0\t1\t7 1\n",
      "86\t6\t0.0\t1\t6\n",
      "87\t1\t0.0\t1\t1\n",
      "88\t10\t0.0\t1\t9 1\n",
      "89\t6\t0.0\t1\t6\n",
      "90\t11\t0.0\t1\t11\n",
      "91\t3\t0.0\t1\t2 1\n",
      "92\t7\t0.0\t1\t7\n",
      "93\t1\t0.0\t1\t1\n",
      "97\t9\t0.0\t1\t7 2\n",
      "98\t11\t0.0\t1\t10 1\n",
      "99\t20\t0.0\t1\t17 3\n",
      "100\t9\t0.0\t1\t6 3\n",
      "101\t5\t0.0\t1\t3 2\n",
      "102\t4\t0.0\t1\t3 1\n",
      "103\t21\t0.0\t1\t18 3\n",
      "104\t3802\t0.0\t1\t3407 379 16\n",
      "105\t16313\t0.0\t1\t14542 1710 61\n",
      "106\t11730\t0.0\t1\t10307 1388 35\n",
      "107\t12051\t0.0\t1\t10550 1432 69\n",
      "108\t6687\t0.0\t1\t5825 815 47\n",
      "109\t5279\t0.0\t1\t4621 625 33\n",
      "110\t111\t0.0\t1\t63 48\n",
      "111\t37\t0.0\t1\t33 4\n",
      "112\t47\t0.0\t1\t43 4\n",
      "113\t23\t0.0\t1\t21 1 1\n",
      "114\t21\t0.0\t1\t20 1\n",
      "115\t13\t0.0\t1\t13\n",
      "116\t7\t0.0\t1\t4 3\n",
      "120\t2\t0.0\t1\t2\n",
      "This is cutadapt 3.5 with Python 3.10.12\n",
      "Command line parameters: -g TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT -j 20 -m 17 -M 23 Veri_L2622.fq -o Trimmed/Veri_L2622.fq\n",
      "Processing reads on 20 cores in single-end mode ...\n",
      "Finished in 0.42 s (2 µs/read; 35.63 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                 249,963\n",
      "Reads with adapters:                   227,656 (91.1%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                 174 (0.1%)\n",
      "Reads that were too long:               52,335 (20.9%)\n",
      "Reads written (passing filters):       197,454 (79.0%)\n",
      "\n",
      "Total basepairs processed:    37,494,450 bp\n",
      "Total written (filtered):      3,987,447 bp (10.6%)\n",
      "\n",
      "=== Adapter 3 ===\n",
      "\n",
      "Sequence: TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT; Type: linked; Length: 18+17; 5' trimmed: 227656 times; 3' trimmed: 227656 times\n",
      "\n",
      "Minimum overlap: 3+3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-18 bp: 1\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-17 bp: 1\n",
      "\n",
      "Overview of removed sequences at 5' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t4\t3905.7\t0\t4\n",
      "5\t1\t244.1\t0\t1\n",
      "12\t2\t0.0\t1\t1 1\n",
      "14\t1\t0.0\t1\t0 0 1\n",
      "16\t3\t0.0\t1\t1 2\n",
      "17\t1\t0.0\t1\t0 1\n",
      "19\t11\t0.0\t1\t9 1 1\n",
      "20\t224\t0.0\t1\t45 167 12\n",
      "21\t27962\t0.0\t1\t24826 3017 119\n",
      "22\t27642\t0.0\t1\t24304 3212 126\n",
      "23\t55906\t0.0\t1\t49342 6365 199\n",
      "24\t45126\t0.0\t1\t39825 5129 172\n",
      "25\t70735\t0.0\t1\t63274 7222 239\n",
      "26\t38\t0.0\t1\t7 31\n",
      "\n",
      "\n",
      "\n",
      "Overview of removed sequences at 3' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t23\t3905.7\t0\t20 3\n",
      "4\t1\t976.4\t0\t1\n",
      "6\t1\t61.0\t0\t1\n",
      "23\t2\t0.0\t1\t2\n",
      "24\t8\t0.0\t1\t6 2\n",
      "25\t11\t0.0\t1\t9 2\n",
      "26\t21\t0.0\t1\t15 6\n",
      "27\t12\t0.0\t1\t8 4\n",
      "28\t7\t0.0\t1\t5 2\n",
      "29\t4\t0.0\t1\t3 1\n",
      "42\t1\t0.0\t1\t0 1\n",
      "46\t1\t0.0\t1\t1\n",
      "47\t1\t0.0\t1\t1\n",
      "48\t1\t0.0\t1\t1\n",
      "49\t1\t0.0\t1\t1\n",
      "50\t1\t0.0\t1\t1\n",
      "54\t1\t0.0\t1\t1\n",
      "55\t1\t0.0\t1\t1\n",
      "56\t2\t0.0\t1\t2\n",
      "57\t2\t0.0\t1\t2\n",
      "58\t2\t0.0\t1\t2\n",
      "59\t1\t0.0\t1\t1\n",
      "62\t3\t0.0\t1\t3\n",
      "63\t414\t0.0\t1\t369 44 1\n",
      "64\t2972\t0.0\t1\t2596 361 15\n",
      "65\t7907\t0.0\t1\t6754 1119 34\n",
      "66\t6087\t0.0\t1\t5154 904 29\n",
      "67\t6120\t0.0\t1\t5100 989 31\n",
      "68\t3513\t0.0\t1\t2945 545 23\n",
      "69\t2415\t0.0\t1\t2027 372 16\n",
      "70\t69\t0.0\t1\t45 21 3\n",
      "71\t23\t0.0\t1\t21 2\n",
      "72\t9\t0.0\t1\t8 1\n",
      "73\t8\t0.0\t1\t7 1\n",
      "74\t14\t0.0\t1\t12 2\n",
      "75\t9\t0.0\t1\t8 1\n",
      "76\t7\t0.0\t1\t6 1\n",
      "77\t7\t0.0\t1\t6 1\n",
      "78\t20\t0.0\t1\t18 2\n",
      "79\t20\t0.0\t1\t16 4\n",
      "80\t19\t0.0\t1\t14 5\n",
      "81\t16\t0.0\t1\t13 3\n",
      "82\t25\t0.0\t1\t24 1\n",
      "83\t17\t0.0\t1\t16 1\n",
      "84\t20\t0.0\t1\t17 3\n",
      "85\t23\t0.0\t1\t19 4\n",
      "86\t7\t0.0\t1\t4 3\n",
      "87\t7\t0.0\t1\t6 1\n",
      "88\t5\t0.0\t1\t5\n",
      "89\t8\t0.0\t1\t7 1\n",
      "90\t11\t0.0\t1\t10 1\n",
      "91\t7\t0.0\t1\t6 1\n",
      "92\t7\t0.0\t1\t7\n",
      "93\t8\t0.0\t1\t6 2\n",
      "94\t5\t0.0\t1\t4 1\n",
      "95\t6\t0.0\t1\t6\n",
      "96\t13\t0.0\t1\t12 1\n",
      "97\t19\t0.0\t1\t18 1\n",
      "98\t20\t0.0\t1\t16 4\n",
      "99\t20\t0.0\t1\t18 2\n",
      "100\t9\t0.0\t1\t9\n",
      "101\t12\t0.0\t1\t10 2\n",
      "102\t19\t0.0\t1\t15 4\n",
      "103\t58\t0.0\t1\t39 18 1\n",
      "104\t12144\t0.0\t1\t10906 1202 36\n",
      "105\t57096\t0.0\t1\t51260 5653 183\n",
      "106\t41007\t0.0\t1\t36470 4388 149\n",
      "107\t43489\t0.0\t1\t38624 4718 147\n",
      "108\t23797\t0.0\t1\t21116 2565 116\n",
      "109\t19553\t0.0\t1\t17478 2005 70\n",
      "110\t300\t0.0\t1\t144 146 10\n",
      "111\t42\t0.0\t1\t38 3 1\n",
      "112\t25\t0.0\t1\t23 2\n",
      "113\t13\t0.0\t1\t9 4\n",
      "114\t16\t0.0\t1\t9 7\n",
      "115\t14\t0.0\t1\t14\n",
      "116\t13\t0.0\t1\t12 0 1\n",
      "117\t19\t0.0\t1\t19\n",
      "118\t12\t0.0\t1\t12\n",
      "119\t15\t0.0\t1\t13 2\n",
      "120\t2\t0.0\t1\t2\n",
      "121\t10\t0.0\t1\t7 1 2\n",
      "122\t2\t0.0\t1\t1 1\n",
      "123\t2\t0.0\t1\t1 1\n",
      "124\t2\t0.0\t1\t1 0 1\n",
      "This is cutadapt 3.5 with Python 3.10.12\n",
      "Command line parameters: -g TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT -j 20 -m 17 -M 23 Veri_L2713.fq -o Trimmed/Veri_L2713.fq\n",
      "Processing reads on 20 cores in single-end mode ...\n",
      "Finished in 0.24 s (2 µs/read; 25.24 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                 101,600\n",
      "Reads with adapters:                    82,430 (81.1%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                 261 (0.3%)\n",
      "Reads that were too long:               33,663 (33.1%)\n",
      "Reads written (passing filters):        67,676 (66.6%)\n",
      "\n",
      "Total basepairs processed:    15,240,000 bp\n",
      "Total written (filtered):      1,369,208 bp (9.0%)\n",
      "\n",
      "=== Adapter 3 ===\n",
      "\n",
      "Sequence: TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT; Type: linked; Length: 18+17; 5' trimmed: 82430 times; 3' trimmed: 82430 times\n",
      "\n",
      "Minimum overlap: 3+3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-18 bp: 1\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-17 bp: 1\n",
      "\n",
      "Overview of removed sequences at 5' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t3\t1587.5\t0\t3\n",
      "7\t1\t6.2\t0\t1\n",
      "10\t1\t0.1\t1\t1\n",
      "12\t1\t0.0\t1\t1\n",
      "15\t1\t0.0\t1\t0 0 1\n",
      "17\t2\t0.0\t1\t2\n",
      "18\t1\t0.0\t1\t1\n",
      "19\t5\t0.0\t1\t3 2\n",
      "20\t101\t0.0\t1\t18 74 9\n",
      "21\t9901\t0.0\t1\t8617 1234 50\n",
      "22\t9851\t0.0\t1\t8509 1283 59\n",
      "23\t20209\t0.0\t1\t17658 2449 102\n",
      "24\t16609\t0.0\t1\t14442 2077 90\n",
      "25\t25719\t0.0\t1\t22720 2917 82\n",
      "26\t25\t0.0\t1\t5 20\n",
      "\n",
      "\n",
      "\n",
      "Overview of removed sequences at 3' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t22\t1587.5\t0\t21 1\n",
      "4\t1\t396.9\t0\t1\n",
      "23\t4\t0.0\t1\t4\n",
      "24\t5\t0.0\t1\t3 2\n",
      "25\t5\t0.0\t1\t3 2\n",
      "26\t4\t0.0\t1\t3 1\n",
      "27\t4\t0.0\t1\t2 2\n",
      "28\t4\t0.0\t1\t4\n",
      "29\t1\t0.0\t1\t1\n",
      "40\t2\t0.0\t1\t1 0 1\n",
      "44\t1\t0.0\t1\t1\n",
      "46\t2\t0.0\t1\t1 1\n",
      "49\t1\t0.0\t1\t1\n",
      "51\t1\t0.0\t1\t1\n",
      "52\t2\t0.0\t1\t2\n",
      "55\t2\t0.0\t1\t2\n",
      "56\t3\t0.0\t1\t3\n",
      "62\t4\t0.0\t1\t4\n",
      "63\t214\t0.0\t1\t190 24\n",
      "64\t1448\t0.0\t1\t1260 183 5\n",
      "65\t3819\t0.0\t1\t3221 580 18\n",
      "66\t3101\t0.0\t1\t2575 498 28\n",
      "67\t2879\t0.0\t1\t2398 455 26\n",
      "68\t1658\t0.0\t1\t1369 271 18\n",
      "69\t1159\t0.0\t1\t958 191 10\n",
      "70\t33\t0.0\t1\t18 14 1\n",
      "71\t2\t0.0\t1\t1 1\n",
      "72\t1\t0.0\t1\t1\n",
      "74\t3\t0.0\t1\t3\n",
      "75\t1\t0.0\t1\t1\n",
      "76\t3\t0.0\t1\t3\n",
      "77\t5\t0.0\t1\t4 1\n",
      "78\t7\t0.0\t1\t6 1\n",
      "79\t4\t0.0\t1\t4\n",
      "81\t3\t0.0\t1\t1 2\n",
      "82\t2\t0.0\t1\t2\n",
      "83\t6\t0.0\t1\t4 1 1\n",
      "84\t8\t0.0\t1\t4 4\n",
      "85\t8\t0.0\t1\t8\n",
      "86\t5\t0.0\t1\t5\n",
      "87\t7\t0.0\t1\t7\n",
      "88\t9\t0.0\t1\t9\n",
      "89\t9\t0.0\t1\t8 1\n",
      "90\t6\t0.0\t1\t6\n",
      "91\t3\t0.0\t1\t3\n",
      "92\t4\t0.0\t1\t4\n",
      "94\t10\t0.0\t1\t9 1\n",
      "95\t2\t0.0\t1\t2\n",
      "96\t3\t0.0\t1\t3\n",
      "97\t3\t0.0\t1\t2 1\n",
      "102\t2\t0.0\t1\t0 2\n",
      "103\t23\t0.0\t1\t15 7 1\n",
      "104\t5234\t0.0\t1\t4619 598 17\n",
      "105\t19137\t0.0\t1\t17019 2053 65\n",
      "106\t14356\t0.0\t1\t12600 1692 64\n",
      "107\t14420\t0.0\t1\t12734 1625 61\n",
      "108\t8111\t0.0\t1\t6994 1072 45\n",
      "109\t6235\t0.0\t1\t5525 684 26\n",
      "110\t174\t0.0\t1\t113 56 5\n",
      "111\t60\t0.0\t1\t54 6\n",
      "112\t28\t0.0\t1\t28\n",
      "113\t15\t0.0\t1\t12 3\n",
      "114\t8\t0.0\t1\t6 1 1\n",
      "117\t7\t0.0\t1\t7\n",
      "118\t11\t0.0\t1\t9 2\n",
      "119\t4\t0.0\t1\t3 1\n",
      "120\t10\t0.0\t1\t10\n",
      "121\t29\t0.0\t1\t24 5\n",
      "122\t25\t0.0\t1\t22 3\n",
      "123\t23\t0.0\t1\t22 1\n",
      "124\t13\t0.0\t1\t9 4\n",
      "125\t12\t0.0\t1\t9 3\n",
      "This is cutadapt 3.5 with Python 3.10.12\n",
      "Command line parameters: -g TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT -j 20 -m 17 -M 23 Veri_L1622.fq -o Trimmed/Veri_L1622.fq\n",
      "Processing reads on 20 cores in single-end mode ...\n",
      "Finished in 0.40 s (2 µs/read; 36.33 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                 241,241\n",
      "Reads with adapters:                   228,468 (94.7%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                 222 (0.1%)\n",
      "Reads that were too long:               43,898 (18.2%)\n",
      "Reads written (passing filters):       197,121 (81.7%)\n",
      "\n",
      "Total basepairs processed:    36,186,150 bp\n",
      "Total written (filtered):      3,981,233 bp (11.0%)\n",
      "\n",
      "=== Adapter 3 ===\n",
      "\n",
      "Sequence: TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT; Type: linked; Length: 18+17; 5' trimmed: 228468 times; 3' trimmed: 228468 times\n",
      "\n",
      "Minimum overlap: 3+3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-18 bp: 1\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-17 bp: 1\n",
      "\n",
      "Overview of removed sequences at 5' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t2\t3769.4\t0\t2\n",
      "9\t1\t0.9\t0\t1\n",
      "11\t1\t0.1\t1\t0 1\n",
      "13\t1\t0.0\t1\t0 1\n",
      "14\t4\t0.0\t1\t1 3\n",
      "15\t5\t0.0\t1\t1 4\n",
      "17\t1\t0.0\t1\t1\n",
      "18\t6\t0.0\t1\t1 4 1\n",
      "19\t7\t0.0\t1\t4 3\n",
      "20\t214\t0.0\t1\t50 155 9\n",
      "21\t27702\t0.0\t1\t24466 3116 120\n",
      "22\t27790\t0.0\t1\t24307 3332 151\n",
      "23\t55790\t0.0\t1\t49016 6528 246\n",
      "24\t45664\t0.0\t1\t40225 5242 197\n",
      "25\t71216\t0.0\t1\t63503 7436 277\n",
      "26\t64\t0.0\t1\t15 45 4\n",
      "\n",
      "\n",
      "\n",
      "Overview of removed sequences at 3' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t9\t3769.4\t0\t8 1\n",
      "6\t2\t58.9\t0\t2\n",
      "23\t1\t0.0\t1\t1\n",
      "24\t10\t0.0\t1\t9 1\n",
      "25\t10\t0.0\t1\t9 1\n",
      "26\t11\t0.0\t1\t9 1 1\n",
      "27\t12\t0.0\t1\t10 2\n",
      "28\t8\t0.0\t1\t7 1\n",
      "29\t5\t0.0\t1\t4 1\n",
      "32\t1\t0.0\t1\t0 1\n",
      "38\t1\t0.0\t1\t1\n",
      "40\t1\t0.0\t1\t1\n",
      "43\t1\t0.0\t1\t1\n",
      "47\t2\t0.0\t1\t2\n",
      "49\t3\t0.0\t1\t3\n",
      "50\t1\t0.0\t1\t1\n",
      "51\t4\t0.0\t1\t3 1\n",
      "52\t2\t0.0\t1\t1 1\n",
      "53\t3\t0.0\t1\t3\n",
      "54\t5\t0.0\t1\t5\n",
      "55\t3\t0.0\t1\t3\n",
      "56\t5\t0.0\t1\t5\n",
      "57\t2\t0.0\t1\t2\n",
      "58\t2\t0.0\t1\t2\n",
      "59\t3\t0.0\t1\t3\n",
      "60\t4\t0.0\t1\t3 1\n",
      "61\t2\t0.0\t1\t1 1\n",
      "62\t5\t0.0\t1\t5\n",
      "63\t466\t0.0\t1\t389 72 5\n",
      "64\t3018\t0.0\t1\t2617 384 17\n",
      "65\t8041\t0.0\t1\t6809 1188 44\n",
      "66\t6510\t0.0\t1\t5475 987 48\n",
      "67\t6357\t0.0\t1\t5333 987 37\n",
      "68\t3643\t0.0\t1\t3067 546 30\n",
      "69\t2456\t0.0\t1\t2049 389 18\n",
      "70\t78\t0.0\t1\t61 16 1\n",
      "71\t12\t0.0\t1\t9 3\n",
      "72\t4\t0.0\t1\t4\n",
      "73\t22\t0.0\t1\t17 5\n",
      "74\t20\t0.0\t1\t16 3 1\n",
      "75\t22\t0.0\t1\t19 3\n",
      "76\t14\t0.0\t1\t12 2\n",
      "77\t13\t0.0\t1\t12 1\n",
      "78\t19\t0.0\t1\t19\n",
      "79\t9\t0.0\t1\t8 1\n",
      "80\t11\t0.0\t1\t10 1\n",
      "81\t6\t0.0\t1\t5 1\n",
      "82\t25\t0.0\t1\t20 5\n",
      "83\t9\t0.0\t1\t6 3\n",
      "84\t14\t0.0\t1\t11 3\n",
      "85\t8\t0.0\t1\t7 1\n",
      "86\t16\t0.0\t1\t15 1\n",
      "87\t14\t0.0\t1\t14\n",
      "88\t19\t0.0\t1\t16 3\n",
      "89\t28\t0.0\t1\t26 2\n",
      "90\t17\t0.0\t1\t13 3 1\n",
      "91\t14\t0.0\t1\t13 1\n",
      "92\t13\t0.0\t1\t13\n",
      "93\t6\t0.0\t1\t6\n",
      "94\t4\t0.0\t1\t4\n",
      "95\t14\t0.0\t1\t11 3\n",
      "96\t7\t0.0\t1\t6 1\n",
      "97\t8\t0.0\t1\t8\n",
      "98\t14\t0.0\t1\t13 1\n",
      "99\t21\t0.0\t1\t18 3\n",
      "100\t9\t0.0\t1\t7 2\n",
      "101\t10\t0.0\t1\t9 1\n",
      "102\t13\t0.0\t1\t12 1\n",
      "103\t40\t0.0\t1\t27 13\n",
      "104\t12568\t0.0\t1\t11271 1248 49\n",
      "105\t56942\t0.0\t1\t51103 5648 191\n",
      "106\t41418\t0.0\t1\t36720 4528 170\n",
      "107\t42752\t0.0\t1\t37739 4814 199\n",
      "108\t23927\t0.0\t1\t21102 2722 103\n",
      "109\t19168\t0.0\t1\t17009 2082 77\n",
      "110\t318\t0.0\t1\t166 142 10\n",
      "111\t58\t0.0\t1\t52 5 1\n",
      "112\t49\t0.0\t1\t36 12 1\n",
      "113\t21\t0.0\t1\t18 3\n",
      "114\t25\t0.0\t1\t21 4\n",
      "115\t14\t0.0\t1\t10 4\n",
      "116\t15\t0.0\t1\t10 5\n",
      "117\t9\t0.0\t1\t7 2\n",
      "118\t7\t0.0\t1\t6 1\n",
      "119\t1\t0.0\t1\t1\n",
      "120\t1\t0.0\t1\t0 1\n",
      "121\t8\t0.0\t1\t7 0 1\n",
      "122\t2\t0.0\t1\t2\n",
      "123\t2\t0.0\t1\t2\n",
      "124\t1\t0.0\t1\t1\n",
      "This is cutadapt 3.5 with Python 3.10.12\n",
      "Command line parameters: -g TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT -j 20 -m 17 -M 23 Veri_plasmid.fq -o Trimmed/Veri_plasmid.fq\n",
      "Processing reads on 20 cores in single-end mode ...\n",
      "Finished in 0.19 s (4 µs/read; 16.33 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  52,869\n",
      "Reads with adapters:                    48,010 (90.8%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:                 379 (0.7%)\n",
      "Reads that were too long:                5,558 (10.5%)\n",
      "Reads written (passing filters):        46,932 (88.8%)\n",
      "\n",
      "Total basepairs processed:     7,930,350 bp\n",
      "Total written (filtered):        952,414 bp (12.0%)\n",
      "\n",
      "=== Adapter 3 ===\n",
      "\n",
      "Sequence: TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT; Type: linked; Length: 18+17; 5' trimmed: 48010 times; 3' trimmed: 48010 times\n",
      "\n",
      "Minimum overlap: 3+3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-18 bp: 1\n",
      "\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-17 bp: 1\n",
      "\n",
      "Overview of removed sequences at 5' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "18\t1\t0.0\t1\t1\n",
      "19\t3\t0.0\t1\t3\n",
      "20\t35\t0.0\t1\t4 31\n",
      "21\t5607\t0.0\t1\t4994 594 19\n",
      "22\t5754\t0.0\t1\t5115 607 32\n",
      "23\t11485\t0.0\t1\t10194 1221 70\n",
      "24\t9313\t0.0\t1\t8456 804 53\n",
      "25\t15805\t0.0\t1\t14514 1251 40\n",
      "26\t7\t0.0\t1\t2 5\n",
      "\n",
      "\n",
      "\n",
      "Overview of removed sequences at 3' end\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t6\t826.1\t0\t5 1\n",
      "4\t4\t206.5\t0\t4\n",
      "5\t1\t51.6\t0\t1\n",
      "26\t1\t0.0\t1\t1\n",
      "58\t1\t0.0\t1\t1\n",
      "63\t10\t0.0\t1\t10\n",
      "64\t42\t0.0\t1\t38 4\n",
      "65\t156\t0.0\t1\t144 12\n",
      "66\t125\t0.0\t1\t119 5 1\n",
      "67\t141\t0.0\t1\t122 19\n",
      "68\t102\t0.0\t1\t95 5 2\n",
      "69\t49\t0.0\t1\t45 4\n",
      "70\t2\t0.0\t1\t0 2\n",
      "71\t2\t0.0\t1\t2\n",
      "72\t1\t0.0\t1\t1\n",
      "73\t4\t0.0\t1\t4\n",
      "74\t1\t0.0\t1\t1\n",
      "81\t1\t0.0\t1\t1\n",
      "86\t4\t0.0\t1\t4\n",
      "87\t5\t0.0\t1\t4 1\n",
      "88\t3\t0.0\t1\t3\n",
      "89\t1\t0.0\t1\t1\n",
      "90\t1\t0.0\t1\t1\n",
      "92\t1\t0.0\t1\t1\n",
      "93\t2\t0.0\t1\t2\n",
      "94\t1\t0.0\t1\t1\n",
      "95\t3\t0.0\t1\t2 1\n",
      "96\t3\t0.0\t1\t3\n",
      "97\t5\t0.0\t1\t4 1\n",
      "98\t1\t0.0\t1\t1\n",
      "100\t3\t0.0\t1\t3\n",
      "101\t5\t0.0\t1\t5\n",
      "102\t9\t0.0\t1\t9\n",
      "103\t41\t0.0\t1\t34 7\n",
      "104\t5089\t0.0\t1\t4641 436 12\n",
      "105\t13054\t0.0\t1\t12001 1011 42\n",
      "106\t9854\t0.0\t1\t8881 913 60\n",
      "107\t9149\t0.0\t1\t8145 954 50\n",
      "108\t5686\t0.0\t1\t5085 571 30\n",
      "109\t3865\t0.0\t1\t3434 420 11\n",
      "110\t220\t0.0\t1\t168 48 4\n",
      "111\t138\t0.0\t1\t108 30\n",
      "112\t73\t0.0\t1\t52 21\n",
      "113\t48\t0.0\t1\t36 12\n",
      "114\t29\t0.0\t1\t20 9\n",
      "115\t24\t0.0\t1\t17 7\n",
      "116\t9\t0.0\t1\t6 3\n",
      "117\t9\t0.0\t1\t6 3\n",
      "118\t7\t0.0\t1\t5 2\n",
      "119\t5\t0.0\t1\t1 4\n",
      "120\t8\t0.0\t1\t3 5\n",
      "121\t1\t0.0\t1\t0 1\n",
      "122\t2\t0.0\t1\t1 1\n",
      "123\t1\t0.0\t1\t1\n",
      "124\t1\t0.0\t1\t0 0 1\n",
      "127\t1\t0.0\t1\t0 0 1\n"
     ]
    }
   ],
   "source": [
    "for file in os.listdir():\n",
    "    if \"Veri_\" in file:\n",
    "        subprocess.call(['cutadapt', '-g', 'TGGAAAGGACGAAACACC...GTTTTAGAGCTAGAAAT', '-j', '20', '-m', '17', '-M', '23',\n",
    "                     file , '-o', \"Trimmed/\" + file])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "67c6b1a2-2fd7-4847-a8fe-3f70ea043566",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/media/ivan/Extreme SSD/Verification ABE Screen/Trimmed')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e6ee77d-494d-412a-b158-00cc1ca91a8d",
   "metadata": {},
   "source": [
    "bowtie-build Verification_sgRNA.fasta Veri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f8b02bfb-9efa-4ce5-98eb-ceb5900b62fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "Veri_samples = ['Veri_L1713', 'Veri_L2622', 'Veri_L2713', 'Veri_L1622', 'Veri_plasmid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e68bdbf0-3458-4b6c-9fe0-5de09c948d6e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "# reads processed: 56001\n",
      "# reads with at least one alignment: 54769 (97.80%)\n",
      "# reads that failed to align: 1232 (2.20%)\n",
      "Reported 54769 alignments\n",
      "# reads processed: 197454\n",
      "# reads with at least one alignment: 194822 (98.67%)\n",
      "# reads that failed to align: 2632 (1.33%)\n",
      "Reported 194822 alignments\n",
      "# reads processed: 67676\n",
      "# reads with at least one alignment: 61852 (91.39%)\n",
      "# reads that failed to align: 5824 (8.61%)\n",
      "Reported 61852 alignments\n",
      "# reads processed: 197121\n",
      "# reads with at least one alignment: 194255 (98.55%)\n",
      "# reads that failed to align: 2866 (1.45%)\n",
      "Reported 194255 alignments\n",
      "# reads processed: 46932\n",
      "# reads with at least one alignment: 34422 (73.34%)\n",
      "# reads that failed to align: 12510 (26.66%)\n",
      "Reported 34422 alignments\n"
     ]
    }
   ],
   "source": [
    "for d in Veri_samples:\n",
    "    subprocess.call(['bowtie', '-n', '2', '-p', '20', '-q', '--norc', '--best', '-l', '17','-x', 'Veri', d + '.fq', d + '.txt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e79069e9-dbb8-4bd4-975c-9cb1ea0dc324",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BCRABL_1776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCRABL_1782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BCRABL_1792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BCRABL_1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BCRABL_1808</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sgName\n",
       "0  BCRABL_1776\n",
       "1  BCRABL_1782\n",
       "2  BCRABL_1792\n",
       "3  BCRABL_1800\n",
       "4  BCRABL_1808"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgNames_fasta = open(\"Verification_sgRNA.fasta\", \"r\").read().splitlines()[::2]\n",
    "sgNames = [sg[1:] for sg in sgNames_fasta]\n",
    "ScreenDF = pd.DataFrame(sgNames, columns=['sgName'])\n",
    "ScreenDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a81e8c5a-f917-49e9-9e58-2f25bda0b11b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Veri_L1713', 54769]\n",
      "['Veri_L2622', 194822]\n",
      "['Veri_L2713', 61852]\n",
      "['Veri_L1622', 194255]\n",
      "['Veri_plasmid', 34422]\n"
     ]
    }
   ],
   "source": [
    "for d in Veri_samples:\n",
    "    Bowtie_output = open((d+\".txt\"), \"r\").read().split('\\t')\n",
    "    sgRNA_IDs = Bowtie_output[2::7] #sgRNA ID list\n",
    "    sg_Frequency = Counter(sgRNA_IDs)\n",
    "    sgCounts = []\n",
    "    for sgN in sgNames:\n",
    "        sgCounts.append(sg_Frequency[sgN])\n",
    "    ScreenDF[d] = sgCounts\n",
    "    print([d, len(sgRNA_IDs)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "559d1f38-00cb-4d05-8787-7b1ce4287d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ScreenDF.to_csv('Veri_sgRNA_Screen_Counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6493844c-20e3-496b-a114-1241769b8cfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "ScreenDF = pd.read_csv('Veri_sgRNA_Screen_Counts.csv').iloc[:, 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "46cfaeb6-e486-4271-8a0e-89d234bc268e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Veri_samples = ['Veri_L1713', 'Veri_L2622', 'Veri_L2713', 'Veri_L1622', 'Veri_plasmid']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4eddbd53-0bb9-4d65-89cb-d3d27b726a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = pd.DataFrame(Veri_samples, columns=['sample'])\n",
    "meta['condition'] = ['D19', 'D0', 'D19', 'D0', 'Plasmid']\n",
    "meta = meta.set_index('sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4c7f1849-845a-44c7-bfb1-846144b52d12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Veri_L1713</th>\n",
       "      <td>D19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Veri_L2622</th>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Veri_L2713</th>\n",
       "      <td>D19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Veri_L1622</th>\n",
       "      <td>D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Veri_plasmid</th>\n",
       "      <td>Plasmid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             condition\n",
       "sample                \n",
       "Veri_L1713         D19\n",
       "Veri_L2622          D0\n",
       "Veri_L2713         D19\n",
       "Veri_L1622          D0\n",
       "Veri_plasmid   Plasmid"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "89b0f300-526b-4ea9-b80b-162ba08a51f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "ScreenDF2 = ScreenDF.set_index('sgName').T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a6d36029-35fc-4a08-b69f-184eb6fc8622",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "/home/ivan/.local/lib/python3.10/site-packages/matplotlib/projections/__init__.py:63: UserWarning: Unable to import Axes3D. This may be due to multiple versions of Matplotlib being installed (e.g. as a system package and as a pip package). As a result, the 3D projection is not available.\n",
      "  warnings.warn(\"Unable to import Axes3D. This may be due to multiple versions of \"\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "/home/ivan/.local/lib/python3.10/site-packages/pydeseq2/dds.py:590: UserWarning: The dispersion trend curve fitting did not converge. Switching to a mean-based dispersion trend.\n",
      "  self._fit_parametric_dispersion_trend()\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "/home/ivan/.local/lib/python3.10/site-packages/pydeseq2/dds.py:448: UserWarning: As the residual degrees of freedom is less than 3, the distribution of log dispersions is especially asymmetric and likely to be poorly estimated by the MAD.\n",
      "  self.fit_dispersion_prior()\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Replacing 0 outlier genes.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "inference = DefaultInference(n_cpus=20)\n",
    "dds = DeseqDataSet(\n",
    "    counts=ScreenDF2,\n",
    "    metadata=meta,\n",
    "    design_factors=\"condition\",\n",
    "    refit_cooks=True,\n",
    "    n_cpus=20)\n",
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0cd0f9d6-7dab-4a8b-b21f-6701ffd14c38",
   "metadata": {},
   "outputs": [],
   "source": [
    "DESeq2_Normed_Counts = pd.DataFrame(dds.layers['normed_counts'], index=ScreenDF2.index, columns=ScreenDF2.columns).T\n",
    "DESeq2_Normed_Freq = DESeq2_Normed_Counts.div(DESeq2_Normed_Counts.sum(axis=0), axis=1)\n",
    "DESeq2_Normed_Freq.to_csv('DESeq2_Normed_Freq_Veri_Screen.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8c6b2c92-7f12-40a1-b20f-426ded1831a3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: condition D19 vs D0\n",
      "                baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
      "sgName                                                                        \n",
      "BCRABL_1776  1735.551038       -0.790591  0.321833  -2.456528  1.402870e-02   \n",
      "BCRABL_1782  3983.107664        1.109251  0.096126  11.539535  8.337423e-31   \n",
      "BCRABL_1792  3406.842440       -0.330971  0.168121  -1.968649  4.899343e-02   \n",
      "BCRABL_1800  2052.655138       -0.291224  0.270226  -1.077705  2.811654e-01   \n",
      "BCRABL_1808  7263.811038        1.142637  0.161960   7.055044  1.725457e-12   \n",
      "BCRABL_1811  2682.252305       -0.729984  0.107051  -6.819001  9.167596e-12   \n",
      "BCRABL_1817  1407.597094       -0.095338  0.097955  -0.973282  3.304132e-01   \n",
      "BCRABL_1824  7211.442386        0.261897  0.069509   3.767811  1.646851e-04   \n",
      "BCRABL_1829  3223.030997       -0.073291  0.246299  -0.297570  7.660313e-01   \n",
      "BCRABL_1831  3333.278727       -0.775409  0.132757  -5.840812  5.194699e-09   \n",
      "BCRABL_1833  7793.468229        1.019696  0.131503   7.754161  8.892925e-15   \n",
      "BCRABL_1838  9346.640076        0.030174  0.087647   0.344271  7.306427e-01   \n",
      "BCRABL_1840  7025.223912        1.072221  0.102017  10.510172  7.755193e-26   \n",
      "BCRABL_1855  1434.487072       -0.641572  0.113039  -5.675680  1.381390e-08   \n",
      "BCRABL_1861  1493.128893        0.527311  0.140289   3.758749  1.707653e-04   \n",
      "BCRABL_1863  3812.312512        0.041488  0.092265   0.449659  6.529562e-01   \n",
      "BCRABL_1869  3758.026816       -0.081810  0.085612  -0.955597  3.392758e-01   \n",
      "BCRABL_1873  2111.974463       -0.772685  0.264885  -2.917059  3.533487e-03   \n",
      "BCRABL_1874  5287.417196       -0.047702  0.143068  -0.333425  7.388133e-01   \n",
      "BCRABL_1879  7541.302290        1.199364  0.092571  12.956146  2.168742e-38   \n",
      "\n",
      "                     padj  \n",
      "sgName                     \n",
      "BCRABL_1776  2.338116e-02  \n",
      "BCRABL_1782  8.337423e-30  \n",
      "BCRABL_1792  7.537451e-02  \n",
      "BCRABL_1800  4.016649e-01  \n",
      "BCRABL_1808  6.901827e-12  \n",
      "BCRABL_1811  3.055865e-11  \n",
      "BCRABL_1817  4.240947e-01  \n",
      "BCRABL_1824  3.415306e-04  \n",
      "BCRABL_1829  7.660313e-01  \n",
      "BCRABL_1831  1.484200e-08  \n",
      "BCRABL_1833  4.446462e-14  \n",
      "BCRABL_1838  7.660313e-01  \n",
      "BCRABL_1840  5.170129e-25  \n",
      "BCRABL_1855  3.453475e-08  \n",
      "BCRABL_1861  3.415306e-04  \n",
      "BCRABL_1863  7.660313e-01  \n",
      "BCRABL_1869  4.240947e-01  \n",
      "BCRABL_1873  6.424522e-03  \n",
      "BCRABL_1874  7.660313e-01  \n",
      "BCRABL_1879  4.337485e-37  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 0.07 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ABE_Veri = DeseqStats(dds, contrast=[\"condition\", \"D19\", \"D0\"])\n",
    "ABE_Veri.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d086d09b-c47c-475e-8eee-7239a72798a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ABE_Veri = ABE_Veri.results_df\n",
    "ABE_Veri['Veri_growthrate'] = np.log(np.power(2, ABE_Veri['log2FoldChange'])*np.exp(0.055*19*24))/(19*24)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f7eda0e2-bc15-49d5-89c4-97e7b41d0245",
   "metadata": {},
   "outputs": [],
   "source": [
    "ABE_Veri['D0_Freq'] = DESeq2_Normed_Freq[['Veri_L1622', 'Veri_L2622']].mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2f6e3794-c0b6-4802-b411-f97035087b56",
   "metadata": {},
   "outputs": [],
   "source": [
    "ABE_Veri.to_csv('ABE_Veri_DESEQ2_output.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "59d551cd-d9c9-49e8-a3d5-6bfa56dfcd60",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ABE_Veri[['D0_Freq', 'Veri_growthrate']].to_csv('ABE_Veri_Screen_Freq_GR240525.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bf6af606-3b7f-46e9-896f-4c610e81ef28",
   "metadata": {},
   "outputs": [],
   "source": [
    "ABE_Veri_DESeq2 = ABE_Veri[['Veri_growthrate', 'D0_Freq','pvalue', 'padj']]\n",
    "ABE_Veri_DESeq2.columns = [['Veri_growthrate', 'Veri_D0_Freq','Veri_pvalue', 'Veri_padj']]\n",
    "ABE_Veri_DESeq2.reset_index().to_csv('ABE_Veri_DESeq2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "05053e13-48b2-4aed-b413-b1319ce2dd0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Veri_growthrate</th>\n",
       "      <th>Veri_D0_Freq</th>\n",
       "      <th>Veri_pvalue</th>\n",
       "      <th>Veri_padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BCRABL_1776</th>\n",
       "      <td>0.053798</td>\n",
       "      <td>0.019475</td>\n",
       "      <td>1.402870e-02</td>\n",
       "      <td>2.338116e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1782</th>\n",
       "      <td>0.056686</td>\n",
       "      <td>0.035626</td>\n",
       "      <td>8.337423e-31</td>\n",
       "      <td>8.337423e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1792</th>\n",
       "      <td>0.054497</td>\n",
       "      <td>0.048364</td>\n",
       "      <td>4.899343e-02</td>\n",
       "      <td>7.537451e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1800</th>\n",
       "      <td>0.054557</td>\n",
       "      <td>0.025172</td>\n",
       "      <td>2.811654e-01</td>\n",
       "      <td>4.016649e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1808</th>\n",
       "      <td>0.056737</td>\n",
       "      <td>0.064944</td>\n",
       "      <td>1.725457e-12</td>\n",
       "      <td>6.901827e-12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1811</th>\n",
       "      <td>0.053890</td>\n",
       "      <td>0.044003</td>\n",
       "      <td>9.167596e-12</td>\n",
       "      <td>3.055865e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1817</th>\n",
       "      <td>0.054855</td>\n",
       "      <td>0.016548</td>\n",
       "      <td>3.304132e-01</td>\n",
       "      <td>4.240947e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1824</th>\n",
       "      <td>0.055398</td>\n",
       "      <td>0.087763</td>\n",
       "      <td>1.646851e-04</td>\n",
       "      <td>3.415306e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1829</th>\n",
       "      <td>0.054889</td>\n",
       "      <td>0.042443</td>\n",
       "      <td>7.660313e-01</td>\n",
       "      <td>7.660313e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1831</th>\n",
       "      <td>0.053821</td>\n",
       "      <td>0.052588</td>\n",
       "      <td>5.194699e-09</td>\n",
       "      <td>1.484200e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1833</th>\n",
       "      <td>0.056550</td>\n",
       "      <td>0.074323</td>\n",
       "      <td>8.892925e-15</td>\n",
       "      <td>4.446462e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1838</th>\n",
       "      <td>0.055046</td>\n",
       "      <td>0.126296</td>\n",
       "      <td>7.306427e-01</td>\n",
       "      <td>7.660313e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1840</th>\n",
       "      <td>0.056630</td>\n",
       "      <td>0.060932</td>\n",
       "      <td>7.755193e-26</td>\n",
       "      <td>5.170129e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1855</th>\n",
       "      <td>0.054025</td>\n",
       "      <td>0.018648</td>\n",
       "      <td>1.381390e-08</td>\n",
       "      <td>3.453475e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1861</th>\n",
       "      <td>0.055802</td>\n",
       "      <td>0.016470</td>\n",
       "      <td>1.707653e-04</td>\n",
       "      <td>3.415306e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1863</th>\n",
       "      <td>0.055063</td>\n",
       "      <td>0.050964</td>\n",
       "      <td>6.529562e-01</td>\n",
       "      <td>7.660313e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1869</th>\n",
       "      <td>0.054876</td>\n",
       "      <td>0.049179</td>\n",
       "      <td>3.392758e-01</td>\n",
       "      <td>4.240947e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1873</th>\n",
       "      <td>0.053825</td>\n",
       "      <td>0.027899</td>\n",
       "      <td>3.533487e-03</td>\n",
       "      <td>6.424522e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1874</th>\n",
       "      <td>0.054927</td>\n",
       "      <td>0.071632</td>\n",
       "      <td>7.388133e-01</td>\n",
       "      <td>7.660313e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCRABL_1879</th>\n",
       "      <td>0.056823</td>\n",
       "      <td>0.066733</td>\n",
       "      <td>2.168742e-38</td>\n",
       "      <td>4.337485e-37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Veri_growthrate Veri_D0_Freq   Veri_pvalue     Veri_padj\n",
       "sgName                                                              \n",
       "BCRABL_1776        0.053798     0.019475  1.402870e-02  2.338116e-02\n",
       "BCRABL_1782        0.056686     0.035626  8.337423e-31  8.337423e-30\n",
       "BCRABL_1792        0.054497     0.048364  4.899343e-02  7.537451e-02\n",
       "BCRABL_1800        0.054557     0.025172  2.811654e-01  4.016649e-01\n",
       "BCRABL_1808        0.056737     0.064944  1.725457e-12  6.901827e-12\n",
       "BCRABL_1811        0.053890     0.044003  9.167596e-12  3.055865e-11\n",
       "BCRABL_1817        0.054855     0.016548  3.304132e-01  4.240947e-01\n",
       "BCRABL_1824        0.055398     0.087763  1.646851e-04  3.415306e-04\n",
       "BCRABL_1829        0.054889     0.042443  7.660313e-01  7.660313e-01\n",
       "BCRABL_1831        0.053821     0.052588  5.194699e-09  1.484200e-08\n",
       "BCRABL_1833        0.056550     0.074323  8.892925e-15  4.446462e-14\n",
       "BCRABL_1838        0.055046     0.126296  7.306427e-01  7.660313e-01\n",
       "BCRABL_1840        0.056630     0.060932  7.755193e-26  5.170129e-25\n",
       "BCRABL_1855        0.054025     0.018648  1.381390e-08  3.453475e-08\n",
       "BCRABL_1861        0.055802     0.016470  1.707653e-04  3.415306e-04\n",
       "BCRABL_1863        0.055063     0.050964  6.529562e-01  7.660313e-01\n",
       "BCRABL_1869        0.054876     0.049179  3.392758e-01  4.240947e-01\n",
       "BCRABL_1873        0.053825     0.027899  3.533487e-03  6.424522e-03\n",
       "BCRABL_1874        0.054927     0.071632  7.388133e-01  7.660313e-01\n",
       "BCRABL_1879        0.056823     0.066733  2.168742e-38  4.337485e-37"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ABE_Veri_DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd519693-d7c0-4062-93fe-caa5b55a63ed",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
